feat(client,challenge-parser): update fill-in-the-blank to support Chinese

huyenltnguyen opened the pull request:
Checklist:

<!-- Please follow this checklist and put an x in each of the boxes, like this: [x]. It will ensure that our team takes your pull request seriously. -->

- [x] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [x] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/how-to-open-a-pull-request/).
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitHub Codespaces.

<!--If your pull request closes a GitHub issue, replace the XXXXX below with the issue number.-->

Sorry, this PR turned out quite large, though half of the changes are tests.

I'll explain the implementation details in the PR comments, but here's the high-level overview:

Chinese fill-in-the blank challenges can have the following forms:
1. Hanzi-Pinyin pair: `我是 BLANK。(wǒ shì BLANK)` 
2. Hanzi only:  `我是 BLANK。`
3. Pinyin only: `wǒ shì BLANK`

The approach is to parse the challenge content, detect any Hanzi–Pinyin pairs, and handle each case accordingly:
- Form 2 and 3 can be handled the same way as non-Chinese challenges.
- For form 1, the parser should detect `BLANK`s only in the Hanzi segment. The `BLANK`s in the Pinyin segment are placeholders/markers, which we use to filter them out in the UI (we don't want to show the pinyin of the missing word).

It's a little cumbersome because we have to perform the check and parsing in two places: challenge parser and client.

---

Closes https://github.com/freeCodeCamp/language-curricula/issues/16

Note: The Chinese fill-in-the-blank challenges may not behave correctly yet, as we still need the new inputs to complete the puzzle. They are tracked separately (https://github.com/freeCodeCamp/language-curricula/issues/14 and https://github.com/freeCodeCamp/language-curricula/issues/18), I'll start working on those right after this one. 

## Steps to test

- Run the app
- Go to `/learn/a1-professional-chinese/zh-a1-learn-understanding-the-questions-and-answers/task-3`

## Screenshots

The text spacing is a little too tight. I'll get to that once I have all Chinese functionality is in place.

<details>
<summary>Screenshot</summary>

<img width="1440" height="950" alt="Screenshot 2025-11-11 at 23 59 04" src="https://github.com/user-attachments/assets/7b49f0ae-6150-43c6-b755-0ad34d171b0d" />

</details>

<!-- Feel free to add any additional description of changes below this line -->

huyenltnguyen committed:
feat(client,challenge-parser): update fill-in-the-blank to support Chinese

huyenltnguyen commented on client/src/templates/Challenges/components/fill-in-the-blanks.tsx:
I think you'd want to review the files in the order of the data flow:
- `add-fill-in-the-blank.js`
- `fill-in-the-blank/show.tsx`
- `components/fill-in-the-blank.tsx`
- `fill-in-the-blank/parse-blanks.ts`

huyenltnguyen commented on tools/challenge-parser/parser/plugins/add-fill-in-the-blank.js:
```diff
@@ -40,19 +42,86 @@ function plugin() {
+
+      file.data.fillInTheBlank = getFillInTheBlank(sentenceNodes, blanksNodes);
+
+      function getFillInTheBlank(sentenceNodes, blanksNodes) {
```
I moved the function inside `transformer()` so it can use `toHtml`, and slightly renamed it (`getfillInTheBlank` to `getFillInTheBlank`).

The function body is the same, except...

huyenltnguyen commented on tools/challenge-parser/parser/plugins/add-fill-in-the-blank.js:
```diff
@@ -40,19 +42,86 @@ function plugin() {
+          throw Error('sentence is missing from fill in the blank');
+        if (!blanks) throw Error('blanks are missing from fill in the blank');
+
+        validateBlankCount(sentenceNodes, blanks);
```
... this.

I moved this check:

https://github.com/freeCodeCamp/freeCodeCamp/blob/3d7dea735baa070c3e36af7c87d2c8e1eb849c74/tools/challenge-parser/parser/plugins/add-fill-in-the-blank.js#L83-L86

into the `validateBlankCount` function. 

I'll explain the function in a comment below.

huyenltnguyen commented on tools/challenge-parser/parser/plugins/add-fill-in-the-blank.js:
```diff
@@ -40,19 +42,86 @@ function plugin() {
+
+            return {
+              answer: blanksNodes[0].children[0].value,
+              feedback: toHtml(feedbackNodes)
```
This is the only change: switching from `mdastToHtml` to `toHtml`.

huyenltnguyen commented on client/src/templates/Challenges/fill-in-the-blank/show.tsx:
```diff
@@ -135,12 +136,26 @@ const ShowFillInTheBlank = ({
+        const { hanzi } = hanziPinyin;
+        // TODO: Implement full hanzi-pinyin validation logic
+        // https://github.com/freeCodeCamp/language-curricula/issues/18
+        return normalizedUserAnswer === hanzi;
```
This is just a placeholder. I'll update the client logic when I add the new inputs in.

huyenltnguyen commented on client/src/templates/Challenges/fill-in-the-blank/parse-blanks.ts:
```diff
@@ -1,6 +1,46 @@
+
+type ParagraphElement = PlainTextNode | HanziPinyinNode | BlankNode;
+
+const HANZI_PINYIN_PATTERN = /^(.+?)\s*\((.+?)\)$/;
```
This is the exact regex we have in the challenge parser:

https://github.com/freeCodeCamp/freeCodeCamp/blob/3dc5cb6f43e232054ebf4ab2f94e5499daabfbff/tools/challenge-parser/parser/plugins/utils/i18n-stringify.js#L9

I realized just now that this regex is pretty naive, that it doesn't validate the characters and assumes anything follows the `text (text)` pattern is Chinese. 

I'll have a follow up PR to fix both this and the one in the challenge parser.

ojeytonwilliams commented on client/src/templates/Challenges/fill-in-the-blank/parse-blanks.test.ts:
```diff
@@ -129,4 +129,131 @@ describe('parseBlanks', () => {
+  it('handles Chinese with multiple BLANKs', () => {
+    expect(
+      parseBlanks('<p>BLANK好，BLANK是王华 (BLANK hǎo BLANK shì Wang Hua)</p>')
+    ).toEqual([
```
I could be missing something, but in general I think we want to be able to do something like this:

`是 (shi) 王 (Wang) 华 (Hua)`

or

`是 (shi) BLANK (BLANK) 华 (Hua)`

rather than having to have all the pinyin together. Sometimes we will want to, but not always, right?

It doesn't seem like the current approach can handle that.

ojeytonwilliams replied:
For instance:

<img width="278" height="98" alt="Image" src="https://github.com/user-attachments/assets/a65f373c-ad62-4c2a-bb10-5c4dc3aba00d" />

from task 3 is tricky to read since the pinyin and hanzi are not aligned.

huyenltnguyen replied:
Yeah, `是 (shi) 王 (Wang) 华 (Hua)` would be a little easier to handle, but I figured the DX wouldn't be ideal since we can't write a natural sentence that way.

The alignment issue is probably due to the spacing between the characters being too tight. I haven't touched any CSS at all, but it should be possible to adjust the spacing.

Here is the prototype: https://huyenltnguyen.github.io/language-playground/chinese-fill-in-the-blank

ojeytonwilliams replied:
My screenreader's just shouting "chinese letter" at me over and over.

I did find some discussion here: https://w3c.github.io/ruby-t2s-req/#a-single-ruby-element-or-multiple-ruby-elements-per-one-compound-word

What I'm getting from that is that it's context dependent where the boundaries go, so authors likely need to be able to group characters according to said context. Typically by compound word, I imagine.

> Multiple ruby / rt would introduce pauses (which is the issue my prototype is having). 

Multiple `ruby`s yes, but not multiple `rt`s. It reads the following (english word) as you'd expect. No long pauses.

<ruby>
  english <rp>(</rp><rt>one</rt><rp>)</rp> word <rp>(</rp><rt>two</rt><rp>)</rp>
</ruby>

However, there's a fun new problem. The second `rt` is ignored by my screen reader. I don't know why, though. As far as I can see, you're allowed to have multiple.

huyenltnguyen committed:
rework the implementation

huyenltnguyen commented:
@ojeytonwilliams I discussed with Miya last week, and we sent this codepen link to a screen reader user for testing: https://codepen.io/huyenltnguyen/pen/WbwwmoQ (basically just a cleaner version of what I shared with you).

In the pen, I presented five options:
1. A single `ruby` for an entire sentence 
2. Multiple `ruby`s and one `rt` each, with the text content split at punctuation
3. A single `ruby` with multiple `rt`s, with the text content split at punctuation
4. A multiple `ruby`s, each contains a single character
5. A single `ruby` with multiple `rt`s, each contains a single character

With:
- Case 1 being equivalent to writing `你好，我是王华 (nǐ hǎo wǒ shì Wang Hua)。`
- Case 2 and 3 being equivalent to writing `你好 (nǐ hǎo)，我是王华 (wǒ shì Wang Hua)。`
- Case 4 and 5 being equivalent to writing `你 (nǐ) 好 (hǎo)，我 (wǒ) 是 (shì) 王 (Wang) 华 (Hua)。`

The user said that case 1 gave the best experience, but case 2 and 3 are also acceptable. I didn't follow up, but I took it that case 2 and 3 behave the same to them, which lines up with my experience (https://github.com/freeCodeCamp/freeCodeCamp/pull/63741#discussion_r2519374386).

I think case 2 and 3 strike a good balance between screen reader experience and UI presentation here. Since multiple `rt`s appear to behave incorrectly on your end / with some screen readers, I went with the multiple `ruby`s - single `rt` approach (case 2).

---

Also regarding when to split, I couldn't find a standard guideline or official recommendation. Most of the recent W3C docs are draft, so they are subject to change.

However, there is a doc that is a bit old (written in 2016) but has some interesting sections:

> Mono vs. group vs. jukugo
>
> Mono ruby makes it easy to handle line breaks when justifying text, since the browser can split the line between any two base characters. It also maps base characters and annotations precisely, and allows styling to apply the fine rendering control you may need.
>
> Group ruby, on the other hand, assigns a single annotation to a sequence of base characters, and these base characters can no longer be split at the end of a line. Situations where group ruby is appropriate include sequences of base characters that are associated with a single phonetic sound, or semantic ruby that applies to a whole word, or even a phrase.

https://www.w3.org/International/articles/ruby/markup#mono_group_etc

> How long should my `ruby` element be?
>
> Given the ability to string multiple ruby pairings together in a single `ruby` element, the question arises as to what is the optimal number of pairings within any given `ruby` element.
>
> You are free to decide this for yourself. If, however, you want to use a jukugo ruby arrangement some time in the future, you will need to establish clear word boundaries, so that annotations don't overlap adjacent words. This may also be important if you want to produce inline versions of your annotations, and you have used the tabular markup approach. That will ensure that the annotations appear after the words that they refer to. In these cases, you should start a new `ruby` element for each word.

https://www.w3.org/International/articles/ruby/markup#words

So basically the takeaway here is... use your judgement.

Since we have to handle internal punctuation marks and even English words, I also found some examples that help guide my approach:
- https://www.w3.org/TR/html-ruby-extensions/#ruby-combine-ex
  - Punctuation marks are left outside of `ruby` element
- https://w3c.github.io/ruby-t2s-req/#examples-of-harmful-double-reading-english
  - English words should not have a ruby annotation

---

So here is my proposal:
- Convention:
  - Group Hanzi with Pinyin using the `hanzi (pinyin)` pattern
  - Split at punctuation marks, English words, `BLANK`, and wherever the author deems appropriate
  - Place the Pinyin before the punctuation mark
  - For fill-in-the-blank sentences, use a single `BLANK` to represent the blank (basically the same as for non-Chinese)
- Examples
  - Static text:
    - `我叫王华 (wǒ jiào Wang Hua)。`
    - `你好 (nǐ hǎo)，我是王华 (wǒ shì Wang Hua)，请问你叫什么名字 (qǐng wèn nǐ jiào shén me míng zi)？`
    - `我是 (wǒ shì) UI 设计师 (shè jì shī)。`
  - Fill-in-the-blank:
    - `BLANK 叫王华 (jiào Wang Hua)。`
    - `BLANK 好 (hǎo)，我 (wǒ) BLANK 王华 (Wang Hua)，BLANK 你叫什么名字 (nǐ jiào shén me míng zi)？`
    - `我 (wǒ) BLANK UI 设计师 (shè jì shī)。`

I'll add a doc for this over /contribute once we're all settled on the approach.

---

As for the implementation, I changed the approach for how the parser handle Chinese fill-in-the-blank:
- Non-Chinese sentences are parsed as plain text and sent to the client for extraction:
    - Input: `BLANK a nice day.`
    - Output: `<p>BLANK a nice day.</p>`
- I previously processed Chinese sentences the same way, and had the client parse the sentence to Hanzi-Pinyin pair:
  - Input: `BLANK 叫王华 (BLANK jiào Wang Hua)。`
  - Output: `<p>BLANK 叫王华 (BLANK jiào Wang Hua)。</p>`
- With multiple groupings, the parsing logic in the client would become ridiculous, so I decided to do that in the parser and send the client the groups as `ruby` elements instead. The extraction on the client side would also be less brittle this way.
  - Input:  `BLANK 叫王华 (jiào Wang Hua)。`
  - Output: `<p> BLANK <ruby>叫王华<rp>(</rp><rt>jiào Wang Hua</rt><rp>)</rp></ruby>。</p>`
---

<details>
<summary>Results of the changes</summary>

<img width="1676" height="977" alt="Screenshot 2025-11-17 at 03 19 44" src="https://github.com/user-attachments/assets/6f92353f-ba89-4ad8-828a-4e68f467fc22" />

<img width="1676" height="979" alt="Screenshot 2025-11-17 at 03 20 04" src="https://github.com/user-attachments/assets/1fe40cb0-5fb3-41cc-ab95-df614f4be61c" />



</details>

huyenltnguyen commented on curriculum/challenges/english/blocks/zh-a1-learn-understanding-the-questions-and-answers/6904b3933a383d68ec0e5f0d.md:
Updated a couple of challenges here so we have some samples to test with. I'll work with Miya to update the rest later.

huyenltnguyen commented on tools/challenge-parser/parser/plugins/add-fill-in-the-blank.js:
```diff
@@ -40,19 +42,102 @@ function plugin() {
+          }
+        }
+
+        return { sentence, blanks, ...(inputType && { inputType }) };
```
Only Chinese fill-in-the-blank challenges can have `inputType`.

ojeytonwilliams reviewed (approved):
Hi Huyen, sorry for the delay.  While this PR did not introduce the bad separation of concerns, it highlights it. The issue being that the parser's main job is to convert the .md files into something that the client can render. However, by creating html strings the parser is deciding some of *how* we should render, rather than just *what* to render.

The broader issue is out of scope, but for this specific task the process ends up being a little more complicated than it needs to be. The process being: parse fillInTheBlank md file -> find the hanzi-pinyin pairs -> transform into js object with html properties -> client -> client parses those html properties to find the hanzi-pinyin pairs again

If the parser provided the hanzi-pinyin data directly to the client, we could skip the client-side parsing entirely.

ojeytonwilliams commented:
> If we want to avoid client-side parsing, IMHO, we should refactor the entire fill-in-the-blank system to use structured data, not just the Chinese-specific part.

I agree, I think it's all or nothing. The extra complexity of the Chinese just highlights the issue.
